feat: Add a red flash on receiving damage

askneller opened the pull request:
### What this PR does

Adds a red flash on the player receiving damage (the death screen background). Overlay stays for 150ms.

![image](https://user-images.githubusercontent.com/17286005/100011145-08b5e600-2e1d-11eb-8204-d0bac1258ce5.png)

### How to Test

Start a game with WildAnimals enabled and spawn in a hostileDeer. You should see the flash when it attacks you.

Requires https://github.com/MovingBlocks/TeraNUI/pull/42

askneller committed:
Add a red flash on receiving damage

skaldarnar commented:
Hej @askneller , thanks a lot for the PR! 🚀 

Some comments/questions even before I do a proper review:

- I think we already have a slight camera shake on hit (but I might be mistaken). Thus, I'm wondering whether we can limit this more drastic feedback to "hard hits", e.g., when the player takes more than X damage in one hit, or looses more than Y% of their health with a single hit. What do you think of that? 
- Many games these days use directional indicators to give a hint to the player where they are damaged from. How complicated would that be here? I could easily provide some assets for this if we want to try it out...

jdrueckert commented:
> or looses more than Y% of their health with a single hit

Maybe Y% of their _current_ health... Thus it will get more frequent and visible the less health the player gets. If you only have 2 hearts left, losing half a heart may already be a hard hit.

askneller commented:
@skaldarnar @jdrueckert  Thanks for the comments. Skaldarnar I can give the directional indicator a try, some assets would be nice thanks.

Jdrueckert how about 25% of current health in one hit?

jdrueckert commented:
> how about 25% of current health in one hit?

Yes I think that sounds like a reasonable value. If during playtests we see that it's too often or maybe not often enough, we can still change that value in the future to improve balancing if necessary.

skaldarnar commented:
@askneller @jdrueckert Here are two variants, one quite simple spanning the whole side, the other slightly triangle shaped only in the center. I can easily export each layer as separate image to have the individual sites separate. Probably a single texture rendered multiple times in the different locations is sufficient, though.

Let me know which one you prefer.

### Line
![terasology_hit-indicator-simple](https://user-images.githubusercontent.com/1448874/100648996-1221df80-3342-11eb-89bd-da83e2b0ba85.jpg)
![hit-indicator-line](https://user-images.githubusercontent.com/1448874/100650198-d9830580-3343-11eb-9453-635e89395557.png)

### Triangle
![terasology_hit-indicator-triangle](https://user-images.githubusercontent.com/1448874/100649007-1817c080-3342-11eb-83cb-67b8542ae6ec.jpg)
![hit-indicator-triangle](https://user-images.githubusercontent.com/1448874/100650204-da1b9c00-3343-11eb-8761-b483235ee935.png)

askneller committed:
Switch to damage direction indication

askneller committed:
Reintroduce major damage overlay

askneller commented:
I have updated the PR so that the flash only appears when the player receives damage >= 25% of their current health. I was also able to add the directional indicator. @skaldarnar I used the triangle one but unfortunately I think the transparency was lost when you pasted it (it has a white background).

Note for the direction to work this requires this PR: https://github.com/Terasology/Behaviors/pull/54

askneller committed:
Merge branch 'develop' of https://github.com/Terasology/Health into damage-overlay

skaldarnar commented:
Thanks a lot @askneller - will check it out today!

skaldarnar commented on src/main/java/org/terasology/logic/health/HealthClientSystem.java:
```diff
@@ -57,51 +62,77 @@ public void initialise() {
-        return rad / Math.PI * 180.0;
+    private double determineDamageDirection(EntityRef instigator, LocationComponent locationComponent) {
+        LocationComponent instigatorLocation = instigator.getComponent(LocationComponent.class);
+        Vector3f loc = new Vector3f();
+        loc = locationComponent.getWorldPosition(loc);
```
These two lines create a new vector `loc` and then write the world position from the component into that vector - so far, so good. 👍

skaldarnar commented on src/main/java/org/terasology/logic/health/HealthClientSystem.java:
```diff
@@ -57,51 +62,77 @@ public void initialise() {
+        loc = locationComponent.getWorldPosition(loc);
+        Vector3f instLoc = new Vector3f();
+        instLoc = instigatorLocation.getWorldPosition(instLoc);
+        Vector3f locDiff = instLoc.sub(loc);
```
Now this line is a bit "tricky" (as in, I think it works in this context, but bears a pitfall I'd like to explain).

`instLoc.sub(loc)` does the following under the hood:

- it mutates the vector you call the method on (`instLoc`) such that it holds the result afterwards
- it returns that vector after mutation (again, `instLoc`) so that you can chain multiple operations

This means that this line **modifies `instLoc`** and also creates a new _alias_ for it under the name `locDiff`.

askneller replied:
Ah, thanks. I guess I was a bit confused regarding using JOML. I'll keep that in mind.

skaldarnar commented on src/main/java/org/terasology/logic/health/HealthClientSystem.java:
```diff
@@ -57,51 +62,77 @@ public void initialise() {
+        LocationComponent instigatorLocation = instigator.getComponent(LocationComponent.class);
+        Vector3f loc = new Vector3f();
+        loc = locationComponent.getWorldPosition(loc);
+        Vector3f instLoc = new Vector3f();
+        instLoc = instigatorLocation.getWorldPosition(instLoc);
+        Vector3f locDiff = instLoc.sub(loc);
+        locDiff.normalize();
```
That said, these four lines could be combined to save one allocation of a new vector. 

```suggestion
		Vector3f locDiff = instigatorLocation.getWorldPosition(new Vector3f()).sub(loc).normalize();
```

skaldarnar commented on src/main/java/org/terasology/logic/health/HealthClientSystem.java:
```diff
@@ -57,51 +62,77 @@ public void initialise() {
+        Vector3f locDiff = instLoc.sub(loc);
+        locDiff.normalize();
+
+        Vector3f worldFacing = new Vector3f();
+        // facing x and z are "how much" of that direction we are facing
+        // e.g. (0.0, 1.0) means that going forward increases world z position without increasing x position
+        worldFacing = locationComponent.getWorldDirection(worldFacing);
+        worldFacing.normalize();
```
This is likely a matter of taste - you _could_ combine this into a single line:

```suggestion
        // facing x and z are "how much" of that direction we are facing
        // e.g. (0.0, 1.0) means that going forward increases world z position without increasing x position
		Vector3f worldFacing = locationComponent.getWorldDirection(new Vector3f()).normalize();
```

askneller committed:
Merge branch 'develop' of https://github.com/Terasology/Health into damage-overlay

askneller committed:
Update src/main/java/org/terasology/logic/health/HealthClientSystem.java

Co-authored-by: Tobias Nett <skaldarnar@googlemail.com>

askneller committed:
Update src/main/java/org/terasology/logic/health/HealthClientSystem.java

Co-authored-by: Tobias Nett <skaldarnar@googlemail.com>

askneller committed:
Checkstyle (Tab to spaces)

jdrueckert committed:
Merge branch 'develop' into damage-overlay

skaldarnar committed:
refactor: make `determineDamageDirection` return a Direction

skaldarnar committed:
chore: update copyright

skaldarnar committed:
fix: make hit indicators truly transparent

skaldarnar committed:
fix: add negative offsets to move hit indicators to edge of screen

skaldarnar committed:
feat: allow to replicate OnDamagedEvent and OnRestoredEvent to clients

skaldarnar committed:
refactor: list required components in event handler signature

skaldarnar commented:
So, testing this in multiplayer revealed that it's not properly working. I've tried to fix it, but this became a bigger issue than expected:

- `OnDamagedEvent` and `OnRestoreEvent` are server-side-only events and never reach the clients - so no way to react on them
- even after adding a default constructor and using `@OwnerEvent` I cannot get the event to be replicated to the clients =/
- in addition, there is this missing type handler (probably not directly related, though)
  ```
  No type handler for type interface com.google.common.collect.SortedSetMultimap used by class org.terasology.logic.health.RegenComponent::regenEndTime
  ```
- the client wants to use the `DelayManager` to perform a delayed action - it seems that the `DelayManager` is not initialized on the client, so the client cannot use it

I don't know how to proceed here, seems to be an impossible task (at least if you want to solve it in the "right" way using the ECS...)

skaldarnar committed:
- remove delay manager
- rename DamageDirection >>> DirectionalDamageOverlay
- handle timing in UI widget

skaldarnar committed:
test: add test for HealthClientSystem#determineDamageDirection

skaldarnar committed:
fix: actually make directional damage indicators work

skaldarnar committed:
remove MajorDamageOverlay (for now)

jdrueckert commented on src/main/java/org/terasology/logic/health/HealthClientSystem.java:
```diff
@@ -1,34 +1,71 @@
+                          HealthComponent healthComponent) {
+        // Show the relevant direction element
+        EntityRef instigator = event.getInstigator();
+        if (instigator != null && instigator.hasComponent(LocationComponent.class)) {
```
We also need to look at the cases in which `instigator` is null, e.g. for falling damage or damage due to Hunger and Thirst.
However, it's fine to do this as a follow-up PR and keep the scope of this PR to "add red flash on receiving damage from other entities"

jdrueckert commented on src/main/java/org/terasology/rendering/nui/layers/hud/DirectionalDamageOverlay.java:
```diff
@@ -0,0 +1,81 @@
+                return Optional.of(left);
+            case BACKWARD:
+                return Optional.of(bottom);
+            case UP:
+            case DOWN:
```
Should these cases return top / bottom? :thinking:

skaldarnar replied:
I couldn't wrap my head around how to best render the indicators if one indicator is used for multiple directions... =/ I think the data structures as they are don't fit that use case.

The concrete problem I had (why the bottom indicator did not show up) was that for `FORWARD`/`UP` the order was "correct", and I first set the image visibility to `false` and then to `true`. For `BACKWARD`/`DOWN` it was the other way around, the second direction canceled out the visibility..

Maybe it would be better to iterate over the images and do the lookup the other way around, and then add some _one-of_ logic to determine whether the indicator is visible...

jdrueckert replied:
Can you write up a follow-up issue for that? Same for the damage indicator for falling and thirst/hunger damage.

skaldarnar replied:
I'd rather directly do the follow-up PR...

jdrueckert commented on src/test/java/org/terasology/logic/health/HealthClientSystemTest.java:
```diff
@@ -0,0 +1,51 @@
+
+        assertEquals(direction, damageDirection);
+    }
+}
```
newline

skaldarnar committed:
style: add newline at end of file
